src/java/main/org/apache/zookeeper/server/ZooKeeperServerMain.java

```diff
@@ -146,14 +146,26 @@ public void runFromConfig(ServerConfig config) throws IOException, AdminServerEx
             );
             containerManager.start();
 
+            // Watch status of ZooKeeper server. If there is an internal error
+            // then will do a graceful shutdown.
+            while (zkServer.isRunning()) {
```

Comment:
This loop works, but I'm not finding it very elegant. In particular, I don't like having a loop + sleep like this to block the thread, so I'm wondering if we can do better than this and have a call that really blocks until the condition is met. For example, we can have a latch that waits until we count down or interrupt it. What do you think?